src/impl/tpls/KokkosSparse_spmv_tpl_spec_decl.hpp

```diff
@@ -295,4 +296,149 @@ namespace Impl {
 } // namespace KokkosSparse
 #endif // KOKKOSKERNELS_ENABLE_TPL_CUSPARSE
 
+#ifdef KOKKOSKERNELS_ENABLE_TPL_MKL
+#include <mkl.h>
+
+namespace KokkosSparse
+{
+namespace Impl
+{
+
+  inline void mkl_safe_call(int errcode)
+  {
+    if(errcode != SPARSE_STATUS_SUCCESS)
+      throw std::runtime_error("MKL returned non-success error code");
+  }
+
+  inline sparse_operation_t mode_kk_to_mkl(char mode_kk)
+  {
+    switch(toupper(mode_kk))
+    {
+      case 'N':
+        return SPARSE_OPERATION_NON_TRANSPOSE;
+      case 'T':
+        return SPARSE_OPERATION_TRANSPOSE;
+      case 'H':
+        return SPARSE_OPERATION_CONJUGATE_TRANSPOSE;
+      default:;
+    }
+    throw std::invalid_argument("Invalid mode for MKL (should be one of N, T, H)");
+  }
+
+  inline void spmv_mkl(sparse_operation_t op, float alpha, float beta,
+      int m, int n, const int* Arowptrs, const int* Aentries, const float* Avalues,
+      const float* x, float* y)
+  {
+    sparse_matrix_t A_mkl;
+    matrix_descr A_descr;
+    A_descr.type = SPARSE_MATRIX_TYPE_GENERAL;
+    A_descr.mode = SPARSE_FILL_MODE_FULL;
+    A_descr.diag = SPARSE_DIAG_NON_UNIT;
+    mkl_safe_call(mkl_sparse_s_create_csr(&A_mkl, SPARSE_INDEX_BASE_ZERO, m, n,
+          const_cast<int*>(Arowptrs), const_cast<int*>(Arowptrs + 1),
+          const_cast<int*>(Aentries), const_cast<float*>(Avalues)));
+    mkl_safe_call(mkl_sparse_s_mv(op, alpha, A_mkl, A_descr, x, beta, y));
```

Comment:
That will not work with intel17 since it uses the old interface which looks like this `mkl_?csrgemv()`
I see that Trilinos is testing with intel 17 so I would be worried about that auto-tester breaking if this is merged there.
The good news is that intel defines in `mkl_version.h` the following variable (showing intel 17 as reference):
`#define INTEL_MKL_VERSION 20170001`

Reply:
I'm glad you knew about this, I had no idea :)

It should be working now for all versions (17, and 18+):
```
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=787 run_time=307
clang-8.0-Pthread_Serial-release build_time=254 run_time=139
clang-9.0.0-Pthread-release build_time=163 run_time=63
clang-9.0.0-Serial-release build_time=166 run_time=52
cuda-10.1-Cuda_OpenMP-release build_time=977 run_time=148
cuda-11.0-Cuda_OpenMP-release build_time=973 run_time=166
gcc-7.3.0-OpenMP-release build_time=173 run_time=49
gcc-7.3.0-Pthread-release build_time=149 run_time=64
gcc-8.3.0-Serial-release build_time=167 run_time=54
gcc-9.1-OpenMP-release build_time=220 run_time=50
gcc-9.1-Serial-release build_time=202 run_time=60
intel-17.0.1-Serial-release build_time=427 run_time=68
intel-18.0.5-OpenMP-release build_time=763 run_time=51
intel-19.0.5-Pthread-release build_time=372 run_time=79
```

Reply:
No worries, when Christian asked about that TPL I also started looking into it too.
I just happened to load the module for intel 17 x)
One cool stuff I also saw is that there is a already a bit of support for OpenMPTarget offload so eventually we will be able to do something clean with MKL in that case!